### DISCUSSION TITLE
how to wrap a large alternation of literals in word boundary assertions and have it be fast?

### DISCUSSION THREAD

**Discussion by Guillermogsjc** [rust-lang/regex#890] [Q&A] [answered]
Hi, to match efficiently large amounts of alternations, I guess it is interesting to trigger `aho_corasick` variant  here https://github.com/rust-lang/regex/blob/9ca3099037dcb2faf1b49e6493f4c758532f2da1/src/exec.rs#L91 regarding doc:

> /// This is only set when the entire regex is a simple unanchored
> /// alternation of literals. We could probably use it more circumstances,
> /// but this is already hacky enough in this architecture.

The question is: is there any way to use word boundaries in such a way this expression is highly optimized for a thing like this?

```
r"\b(a|... #massive ammount of literal alternations here# ...|z)\b"
```
or with `(?-u:\b)` instead of `\b` .

And... regarding PERFORMANCE documentation [here](https://github.com/rust-lang/regex/blob/master/PERFORMANCE.md#resist-the-temptation-to-optimize-regexes)

> there is no problem with using non-greedy matching **or having lots of alternations in your regex**

this previously stated regex would be in the set of "no problem" ?

Thanks

**Comment by BurntSushi** [comment, marked as answer]
I've turned this into a feature request in #891. 



> And... regarding PERFORMANCE documentation [here](https://github.com/rust-lang/regex/blob/master/PERFORMANCE.md#resist-the-temptation-to-optimize-regexes)
> 
> > there is no problem with using non-greedy matching **or having lots of alternations in your regex**
> 
> this previously stated regex would be in the set of "no problem" ?

Yes. But that doesn't mean it gets the Aho-Corasick optimization. The section you're quoting is really written relative to backtracking engines. Namely, a backtracking engine will (absent other optimizations) try each alternate one after the other. This will wind up being slower than even the slowest regex engine in this crate (the PikeVM).

I am being somewhat hand wavy here. Being precise here without concrete examples in front of us is difficult.
